File: languages/csharp/AppEncryption/AppEncryption.Tests/Crypto/CryptoPolicyTest.cs
using System;
using GoDaddy.Asherah.Crypto;
using GoDaddy.Asherah.Crypto.ExtensionMethods;
using Moq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.Crypto
{
    [Collection("Logger Fixture collection")]
    public class CryptoPolicyTest
    {
        private readonly Mock<CryptoPolicy> cryptoPolicyMock;

        public CryptoPolicyTest()
        {
            cryptoPolicyMock = new Mock<CryptoPolicy>();
        }

        [Fact]
        private void IsInlineKeyRotation()
        {
            const CryptoPolicy.KeyRotationStrategy expectedStrategy = CryptoPolicy.KeyRotationStrategy.Inline;
            cryptoPolicyMock.Setup(x => x.GetKeyRotationStrategy()).Returns(expectedStrategy);
            cryptoPolicyMock.Setup(x => x.IsInlineKeyRotation()).CallBase();
            cryptoPolicyMock.Setup(x => x.IsQueuedKeyRotation()).CallBase();

            Assert.True(cryptoPolicyMock.Object.IsInlineKeyRotation());
            Assert.False(cryptoPolicyMock.Object.IsQueuedKeyRotation());
        }

        [Fact]
        private void IsQueuedKeyRotation()
        {
            const CryptoPolicy.KeyRotationStrategy expectedStrategy = CryptoPolicy.KeyRotationStrategy.Queued;
            cryptoPolicyMock.Setup(x => x.GetKeyRotationStrategy()).Returns(expectedStrategy);
            cryptoPolicyMock.Setup(x => x.IsInlineKeyRotation()).CallBase();
            cryptoPolicyMock.Setup(x => x.IsQueuedKeyRotation()).CallBase();

            Assert.True(cryptoPolicyMock.Object.IsQueuedKeyRotation());
            Assert.False(cryptoPolicyMock.Object.IsInlineKeyRotation());
        }

        [Fact]
        private void TestDefaultSystemKeyPrecision()
        {
            cryptoPolicyMock.Setup(x => x.TruncateToSystemKeyPrecision(It.IsAny<DateTimeOffset>())).CallBase();

            DateTimeOffset now = DateTimeOffset.UtcNow;
            DateTimeOffset actualTime = cryptoPolicyMock.Object.TruncateToSystemKeyPrecision(now);
            Assert.Equal(now.Truncate(TimeSpan.FromMinutes(1)), actualTime);
        }

        [Fact]
        private void TestDefaultIntermediateKeyPrecision()
        {
            cryptoPolicyMock.Setup(x => x.TruncateToIntermediateKeyPrecision(It.IsAny<DateTimeOffset>())).CallBase();

            DateTimeOffset now = DateTimeOffset.UtcNow;
            DateTimeOffset actualTime = cryptoPolicyMock.Object.TruncateToIntermediateKeyPrecision(now);
            Assert.Equal(now.Truncate(TimeSpan.FromMinutes(1)), actualTime);
        }
    }
}

File: java/app-encryption/src/test/java/com/godaddy/asherah/crypto/CryptoPolicyTest.java
package com.godaddy.asherah.crypto;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class CryptoPolicyTest {
  @Mock
  CryptoPolicy cryptoPolicy;

  @Test
  void isInlineKeyRotation() {
    CryptoPolicy.KeyRotationStrategy expectedStrategy = CryptoPolicy.KeyRotationStrategy.INLINE;
    when(cryptoPolicy.keyRotationStrategy()).thenReturn(expectedStrategy);
    when(cryptoPolicy.isInlineKeyRotation()).thenCallRealMethod();
    when(cryptoPolicy.isQueuedKeyRotation()).thenCallRealMethod();

    assertTrue(cryptoPolicy.isInlineKeyRotation());
    assertFalse(cryptoPolicy.isQueuedKeyRotation());
  }

  @Test
  void isQueuedKeyRotation() {
    CryptoPolicy.KeyRotationStrategy expectedStrategy = CryptoPolicy.KeyRotationStrategy.QUEUED;
    when(cryptoPolicy.keyRotationStrategy()).thenReturn(expectedStrategy);
    when(cryptoPolicy.isQueuedKeyRotation()).thenCallRealMethod();
    when(cryptoPolicy.isInlineKeyRotation()).thenCallRealMethod();

    assertTrue(cryptoPolicy.isQueuedKeyRotation());
    assertFalse(cryptoPolicy.isInlineKeyRotation());
  }

  @Test
  void testDefaultSystemKeyPrecision() {
    when(cryptoPolicy.truncateToSystemKeyPrecision(any())).thenCallRealMethod();

    Instant now = Instant.now();
    assertEquals(now.truncatedTo(ChronoUnit.MINUTES), cryptoPolicy.truncateToSystemKeyPrecision(now));
  }

  @Test
  void testDefaultIntermediateKeyPrecision() {
    when(cryptoPolicy.truncateToIntermediateKeyPrecision(any())).thenCallRealMethod();

    Instant now = Instant.now();
    assertEquals(now.truncatedTo(ChronoUnit.MINUTES), cryptoPolicy.truncateToIntermediateKeyPrecision(now));
  }

}
